flink-container/src/main/java/org/apache/flink/container/entrypoint/ClassPathJobGraphRetriever.java

```diff
@@ -83,10 +98,38 @@ public JobGraph retrieveJobGraph(Configuration configuration) throws FlinkExcept
 
 	private PackagedProgram createPackagedProgram() throws FlinkException {
 		try {
-			final Class<?> mainClass = getClass().getClassLoader().loadClass(jobClassName);
+			final Class<?> mainClass = getClass().getClassLoader().loadClass(findEntryClass());
 			return new PackagedProgram(mainClass, programArguments);
-		} catch (ClassNotFoundException | ProgramInvocationException e) {
+		} catch (ClassNotFoundException | IOException | ProgramInvocationException e) {
 			throw new FlinkException("Could not load the provided entrypoint class.", e);
 		}
 	}
+
+	private String findEntryClass() throws IOException {
+		if (startsWithFromJarPrefix(jobClassName)) {
+			File jarFile = getJarFile(jobClassName);
+			LOG.info("Parsing manifest of '{}' to find job entry class", jarFile.getAbsolutePath());
+
+			// We find the entry class manually here instead of providing the JAR to PackagedProgram, because
+			// we want the exact same behavior in all cases, i.e. when providing a JAR or a job class name.
+			//
+			// Using PackagedProgram with a JAR as argument would have slightly different behavior than the code
```

Comment:
what is the difference?

Reply:
Like building a user code class loaders whereas the `PackagedProgramUtils` explicitly don't do it. I'll update the comment accordingly.